src/src/com/microsoft/aad/adal/StorageHelper.java

```diff
@@ -305,71 +280,204 @@ public String decrypt(String value) throws GeneralSecurityException, IOException
         Logger.v(TAG, "Finished decryption");
         return decrypted;
     }
+    
+    /**
+     * Get Secret Key based on API level to use in encryption. Decryption key
+     * depends on version# since user can migrate to new Android.OS
+     *
+     * @return
+     * @throws NoSuchAlgorithmException
+     */
+    public synchronized SecretKey loadSecretKeyForEncryption() throws IOException, GeneralSecurityException {
+        // Loading key only once for performance. If API is upgraded, it will
+        // restart the device anyway. It will load the correct key for new API.
+        if (mKey != null && mMacKey != null)
+            return mKey;
 
-    private char getEncodeVersionLengthPrefix() {
-        return (char)('a' + ENCODE_VERSION.length());
+        final byte[] secretKeyData = AuthenticationSettings.INSTANCE.getSecretKeyData();
+        if(secretKeyData == null && Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN_MR2) {
+            throw new IllegalStateException("Secret key must be provided for API < 18. " +
+                    "Use AuthenticationSettings.INSTANCE.setSecretKey()");
+        }
+
+        if (secretKeyData != null) {
+            Logger.v(TAG, "Encryption will use user provided secret key.");
+            mKey = getSecretKey(secretKeyData);
+            mMacKey = getMacKey(mKey);
+            mBlobVersion = VERSION_USER_DEFINED;
+        } else {
+            // androidKeyStore can store app specific self signed cert.
+            // Asymmetric cryptography is used to protect the session
+            // key for Encryption and HMac.
+            // If user specifies secret key, it will use the provided
+            // key.
+            Logger.v(TAG, "Encryption will use keys persisted in AndroidKeyStore.");
+            mKey = getSecretKeyFromAndroidKeyStore();
+            mMacKey = getMacKey(mKey);
+            mBlobVersion = VERSION_ANDROID_KEY_STORE;
+        }
+
+        return mKey;
     }
+    
+    /**
+     * load key based on version for migration
+     * 
+     * @throws GeneralSecurityException
+     * @throws IOException
+     */
+    public synchronized SecretKey loadSecretKeyForDecryption(String keyVersion) throws GeneralSecurityException, IOException {
+
+        if (keyVersion.equals(VERSION_USER_DEFINED)) {
+            return getSecretKey(AuthenticationSettings.INSTANCE.getSecretKeyData());
+        }
 
+        if (keyVersion.equals(VERSION_ANDROID_KEY_STORE)) {
+            if (Build.VERSION.SDK_INT >= 18) {
+                // androidKeyStore can store app specific self signed cert.
+                // Asymmetric cryptography is used to protect the session key
+                // used for Encryption and HMac
+                return readSecretKeyFromKeyStoreForDecryption();
+            } else {
+                throw new IllegalArgumentException(
```

Comment:
Why not throw this:

   +            throw new IllegalStateException("Secret key must be provided for API < 18. " +  
 300   +                    "Use AuthenticationSettings.INSTANCE.setSecretKey()");